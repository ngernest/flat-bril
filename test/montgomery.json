{
  "functions": [
    {
      "args": [
        {
          "name": "r",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "nn",
          "type": "int"
        },
        {
          "name": "t",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "t",
            "r"
          ],
          "dest": "tr",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tr",
            "nn"
          ],
          "dest": "trnn",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "trnn",
            "r"
          ],
          "dest": "m",
          "funcs": [
            "mod"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "m",
            "n"
          ],
          "dest": "mn",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "t",
            "mn"
          ],
          "dest": "tmn",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "tmn",
            "r"
          ],
          "dest": "t",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "t",
            "n"
          ],
          "dest": "b",
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "b"
          ],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "t",
            "n"
          ],
          "dest": "res",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "res"
          ],
          "op": "print"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "t"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "q",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "y",
            "q"
          ],
          "dest": "xq",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "xq"
          ],
          "dest": "result",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
