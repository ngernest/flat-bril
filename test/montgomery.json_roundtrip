{
  "functions": [
    {
      "args": [
        {
          "name": "r",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "nn",
          "type": "int"
        },
        {
          "name": "t",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "t",
            "r"
          ],
          "dest": "tr",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tr",
            "nn"
          ],
          "dest": "trnn",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "trnn",
            "r"
          ],
          "dest": "m",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "m",
            "n"
          ],
          "dest": "mn",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "t",
            "mn"
          ],
          "dest": "tmn",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "tmn",
            "r"
          ],
          "dest": "t",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "t",
            "n"
          ],
          "dest": "b",
          "funcs": [],
          "labels": [],
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "b"
          ],
          "funcs": [],
          "labels": [
            "then",
            "else"
          ],
          "op": "br"
        },
        {
          "label": "then"
        },
        {
          "args": [
            "t",
            "n"
          ],
          "dest": "res",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "res"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "label": "else"
        },
        {
          "args": [
            "t"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "q",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "y",
            "q"
          ],
          "dest": "xq",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "xq"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
