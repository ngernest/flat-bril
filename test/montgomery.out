@main
	op: call 		dest: tr		type: int		args: ["t", "r"]		funcs: ["mod"]
	op: mul  		dest: trnn		type: int		args: ["tr", "nn"]	
	op: call 		dest: m 		type: int		args: ["trnn", "r"]		funcs: ["mod"]
	op: mul  		dest: mn		type: int		args: ["m", "n"]	
	op: add  		dest: tmn		type: int		args: ["t", "mn"]	
	op: div  		dest: t 		type: int		args: ["tmn", "r"]	
	op: ge   		dest: b 		type: bool		args: ["t", "n"]	
	op: br   		args: ["b"]		labels: ["then", "else"]
.then
	op: sub  		dest: res		type: int		args: ["t", "n"]	
	op: print		args: ["res"]	
.else
	op: print		args: ["t"]	
@mod
	op: div  		dest: q 		type: int		args: ["x", "y"]	
	op: mul  		dest: xq		type: int		args: ["y", "q"]	
	op: sub  		dest: result		type: int		args: ["x", "xq"]	
	op: ret  		args: ["result"]	
